eva/executor/create_index_executor.py

```diff
@@ -56,40 +55,32 @@ def exec(self):
                 col for col in feat_catalog_entry.columns if col.name == feat_col_name
             ][0]
 
+            # Get udf function.
+            udf_func = self.node.udf_func
+
             # Add features to index.
             # TODO: batch size is hardcoded for now.
             index = None
-            input_dim, num_rows = -1, 0
-            logical_to_row_id = dict()
+            input_dim = -1
             storage_engine = StorageEngine.factory(feat_catalog_entry)
-            for batch in storage_engine.read(feat_catalog_entry, 1):
-                # Pandas wraps numpy array as an object inside a numpy
-                # array. Use zero index to get the actual numpy array.
-                feat = batch.column_as_numpy_array(feat_col_name)[0]
+            for input_batch in storage_engine.read(feat_catalog_entry, 1):
+                if udf_func:
+                    input_batch.modify_column_alias(feat_catalog_entry.name.lower())
+                    feat_batch = self.node.udf_func.evaluate(input_batch)
+                    feat_batch.drop_column_alias()
+                    input_batch.drop_column_alias()
+                    feat = feat_batch.column_as_numpy_array("features")[0]
+                else:
+                    # Pandas wraps numpy array as an object inside a numpy
+                    # array. Use zero index to get the actual numpy array.
+                    feat = input_batch.column_as_numpy_array(feat_col_name)[0]
                 if index is None:
                     input_dim = feat.shape[-1]
                     index = self._create_index(self.node.index_type, input_dim)
-                index.add(feat)
-
-                # Secondary mapping.
-                row_id = batch.column_as_numpy_array(IDENTIFIER_COLUMN)[0]
-                logical_to_row_id[num_rows] = row_id
 
-                num_rows += 1
-
-            # Build secondary index maaping.
-            secondary_index_catalog_entry = self._create_secondary_index_table()
-            storage_engine = StorageEngine.factory(secondary_index_catalog_entry)
-            storage_engine.create(secondary_index_catalog_entry)
-
-            # Write mapping.
-            logical_id, row_id = list(logical_to_row_id.keys()), list(
-                logical_to_row_id.values()
-            )
-            secondary_index = Batch(
-                pd.DataFrame(data={"logical_id": logical_id, "row_id": row_id})
-            )
-            storage_engine.write(secondary_index_catalog_entry, secondary_index)
+                # Row ID for mapping back to the row.
+                row_id = input_batch.column_as_numpy_array(IDENTIFIER_COLUMN)[0]
+                index.add_with_ids(feat, np.array([row_id]))
 
             # Persist index.
             faiss.write_index(index, self._get_index_save_path())
```

Comment:
This are all `faiss`-specific implementation. I think it will be better to check the `index_type` at the begining of the executor instead of in the `_create_index`.

Reply:
Sure. Refactor the code.